### PR TITLE
Update binding documentation for threadsafe IO

### PR DESCRIPTION
**Pull request by rcurtin**
This is a follow-up to #2995 which has its own issue: #3049.

Basically, when the bindings were revamped to be threadsafe, I forgot to update the guide that describes the automatic binding system.  So, this PR just updates everything, and so hopefully it should be an easier (and more up-to-date!) read for anyone interested in learning about the binding system. :)

### CODE REVIEW

**Review by shrit — APPROVED** [review]
Looks good to me, nothing from my side :+1:

File: doc/guide/bindings.hpp [hunk 1]
```diff
@@ -1214,9 +1285,8 @@ binding.
 
 @subsection bindings_python_testing Testing the Python bindings
 
-We cannot do our tests only from the Boost Unit Test Framework in C++ because we
-need to see that we are able to load parameters properly from Python and return
-output correctly.
+We cannot do our tests only from the Catch in C++ because we need to see that we
+are able to load parameters properly from Python and return output correctly.
```
Comment (zoq) [inline doc/guide/bindings.hpp:1289]: I think this is difficult to read, maybe just remove this section and in the next section, mention that besides the unit tests written with the Catch framework we also have auxiliary Python tests to test the specific binding functionality?

**Comment by zoq** [comment]
let's keep this open

File: doc/guide/bindings.hpp (hunk 1 above)
Comment (rcurtin) [reply doc/guide/bindings.hpp]: Sorry it took me so long to get back to this.  I tried updating this bit in 5fb3ed04e9; what do you think?

**Pull request merged**